### PR TITLE
Improve store to not load chunks from obj store for Series API

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Using the similar idea of https://github.com/prometheus/prometheus/pull/8050 to only fetch labels for `/api/v1/series` API. Changes:

1. Only sort and dedup labels when there is at least one chunk in the specified time range. If no chunk is valid, then not necessary to sort labels.

2. Series query not load chunks.

### CODE REVIEW

**Comment by pracucci** [comment]
@gouthamve FYI, given you're working on this on Cortex side too.

File: pkg/store/bucket.go
Comment (pracucci) [inline pkg/store/bucket.go:778]: Do we need to duplicate `blockSeries()`? How more dirty would be `blockSeries()` code if we add a "skipChunks" parameter to that function? Or could we make it implicitly skipping chunks if `chunkr` is `nil`?

**Review by pracucci — COMMENTED** [review]
Thanks for working on it! I have a design question.

File: pkg/store/bucket.go
Comment (yeya24) [reply pkg/store/bucket.go:778]: Thanks, let me try

Comment (yeya24) [reply pkg/store/bucket.go]: Done. Please take a look!

**Comment by yeya24** [comment]
Tested this pr using our tool https://github.com/observatorium/up to send `/api/v1/series` request.

Red line is master and green line is this pr. Seems this is faster.

Screenshot from 2020-10-19 11-59-34

**Review by GiedriusS — APPROVED** [review]
Looks good to me! Much cleaner now that it is in the same function, IMHO. But let's wait for another review because this is a relatively critical path

File: pkg/store/bucket.go
Comment (kakkoyun) [inline pkg/store/bucket.go:717]: Should this be named `skipped` or `found`? 
```suggestion
		var skipped bool
```
The comment says:
> find is used to check whether there is at least one chunk in the required time range.

But it's only set the `true` when it's skipped. I'm a bit confused here about the naming. Could we be more clear here?

File: pkg/store/bucket.go
Comment (kakkoyun) [inline pkg/store/bucket.go:743]: I think moving this outside the enclosing `if` could make it more readable.

**Review by kakkoyun — CHANGES_REQUESTED** [review]
Looks great. I have only a couple of nits.

File: pkg/store/bucket.go
Comment (yeya24) [reply pkg/store/bucket.go:743]: Thanks. This makes sense!

File: pkg/store/bucket.go
Comment (yeya24) [reply pkg/store/bucket.go]: Done

File: pkg/store/bucket.go
Comment (kakkoyun) [inline pkg/store/bucket.go:744]: How about setting hasValidChunk to true in the clause and just checking it in below `if`? Would this work?

Comment (yeya24) [reply pkg/store/bucket.go:745]: Done.

**Review by kakkoyun — APPROVED** [review]
💯 🥇 💜

**Pull request merged**
